mountpoint-s3/tests/reftests/harness.rs

```diff
@@ -23,14 +55,122 @@ pub struct Harness {
 }
 
 impl Harness {
-    fn new(fs: S3Filesystem<Arc<MockClient>, ThreadPool>, reference: Reference, readdir_limit: usize) -> Self {
+    /// Create a new test harness
+    pub fn new(fs: S3Filesystem<Arc<MockClient>, ThreadPool>, reference: Reference, readdir_limit: usize) -> Self {
         Self {
             readdir_limit,
             reference,
             fs,
         }
     }
 
+    /// Run a sequence of mutation operations on the test harness, checking equivalence between the
+    /// reference model and file system after each operation.
+    pub async fn run(&mut self, ops: Vec<Op>) {
+        for op in ops {
+            debug!(?op, "executing operation");
+            match &op {
+                Op::WriteFile(name, directory_index, contents) => {
+                    let dir = Path::new(directory_index.get(&self.reference));
+                    let full_path = format!("{}/{name}", dir.display());
```

Comment:
Is there a reason to prefer this to `dir.join(name)...`?

Reply:
Ended up doing this and then changing `add_file` and `lookup` to take `impl AsRef<Path>` instead of `&str`, so that we're not doing manual path manipulation anywhere.

Reply:
Nice, I like it